### PR TITLE
refactor: Treat CDataStream bytes as uint8_t

### PR DESCRIPTION
**Pull request by maflcko**
Using `uint8_t` for raw bytes has a style benefit:
* The signedness is clear from reading the code, as it does not depend on the architecture

Other clean-ups in this pull include:
* Remove unused methods
* Constructor is simplified with `Span`
* Remove `Init()` member in favor of C++11 member initialization

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK: nice cleanup!

**Comment by laanwj** [comment]
Concept ACK!

**Comment by laanwj** [comment]
There are some more unnecessary `MakeUCharSpan` that can be removed here, especially related to Base64 handling:
```
src/rpc/rawtransaction.cpp:    return EncodeBase64(MakeUCharSpan(ssTx));
src/rpc/rawtransaction.cpp:    return EncodeBase64(MakeUCharSpan(ssTx));
src/rpc/rawtransaction.cpp:    return EncodeBase64(MakeUCharSpan(ssTx));
src/rpc/rawtransaction.cpp:    return EncodeBase64(MakeUCharSpan(ssTx));
src/rpc/rawtransaction.cpp:    return EncodeBase64(MakeUCharSpan(ssTx));
```

**Comment by jonatack** [comment]
Concept ACK

File: src/streams.h [hunk 1]
```diff
@@ -221,60 +222,25 @@ class CDataStream
...
     CDataStream(int nTypeIn, int nVersionIn, Args&&... args)
+        : nType{nTypeIn},
+          nVersion{nVersionIn}
     {
-        Init(nTypeIn, nVersionIn);
         ::SerializeMany(*this, std::forward<Args>(args)...);
     }
 
-    void Init(int nTypeIn, int nVersionIn)
+    void Rewind()
```
Comment (laanwj) [inline src/streams.h:241]: There is already a `bool Rewind(size_type n)`, overloading it here is confusing. I we should think of a new name. I had hoped to think of a way to use the existing method but it appears there is  no way to access `nReadPos` from the outside.

Comment (MarcoFalke) [reply src/streams.h:241]: I think the name make sense. `Rewind(size_t)` rewinds only how much is passed. `Rewind()` rewinds all the way. Though, I am happy to rename this to anything else, if there are suggestions.

**Comment by laanwj** [comment]
Code review ACK apart from the above comment. This PR is straightforward to review. The first commit can be seen separately.

File: src/streams.h (hunk 1 above)
Comment (theStack) [reply src/streams.h:241]: It's a pity that there is no defined invalid value for size_t, which could be passed as default value. Would `Rewind(std::optional<size_t> n = std::nullopt)` be a possibility here to unite the two methods? Like e.g.:

```diff
diff --git a/src/streams.h b/src/streams.h
index 2a921c2c5..a6e781b95 100644
--- a/src/streams.h
+++ b/src/streams.h
@@ -14,6 +14,7 @@
 #include <assert.h>
 #include <ios>
 #include <limits>
+#include <optional>
 #include <stdint.h>
 #include <stdio.h>
 #include <string.h>
@@ -238,11 +239,6 @@ public:
         ::SerializeMany(*this, std::forward<Args>(args)...);
     }

-    void Rewind()
-    {
-        nReadPos = 0;
-    }
-
     std::string str() const
     {
         return (std::string(begin(), end()));
@@ -339,12 +335,17 @@ public:
         nReadPos = 0;
     }

-    bool Rewind(size_type n)
+    bool Rewind(std::optional<size_type> n = std::nullopt)
     {
+        // Total rewind if no size is passed
+        if (!n) {
+            nReadPos = 0;
+            return true;
+        }
         // Rewind by n characters if the buffer hasn't been compacted yet
-        if (n > nReadPos)
+        if (*n > nReadPos)
             return false;
-        nReadPos -= n;
+        nReadPos -= *n;
         return true;
     }
```

Probably overkill for  methods that are only ever called in the benchmarks though...

**Comment by theStack** [comment]
Concept ACK -- will review as soon as https://github.com/bitcoin/bitcoin/pull/20464#discussion_r546231371 is resolved

File: src/streams.h (hunk 1 above)
Comment (MarcoFalke) [reply src/streams.h]: Thanks, fixed

**Review by theStack — APPROVED** [review]
ACK fa292724598c273867bc6dbf311f1440fe2541ba 🍾

File: src/streams.h (hunk 1 above)
Comment (laanwj) [reply src/streams.h]: > I think the name make sense. Rewind(size_t) rewinds only how much is passed.

Right. My remark was triggered mostly by that the two implementations of `rewind()` were in different places in the method list and not together, the name makes some sense.

But I like the new implementation.

**Comment by laanwj** [comment]
code review ACK fa292724598c273867bc6dbf311f1440fe2541ba

**Pull request merged**